src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,166 @@
+            }
+        }
+        size++;
+        elementArray = (T[]) defaultNewCapacity;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        if (list.size() > size || list.size() < 0) {
+            throw new ArrayListIndexOutOfBoundsException("list " + list.size()
+                    + " out of bounds for length " + size);
+        }
+        for (int i = 0; i < list.size(); i++) {
+            elementArray[size + i] = list.get(i);
+        }
+        size = size + list.size();
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (index >= size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Index " + index
+                    + " out of bounds for length " + elementArray.length);
+        }
+        return elementArray[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        if (index >= size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Index " + index
+                    + " out of bounds for length " + size);
+        }
+        elementArray[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        if (index >= size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Index " + index
+                    + " out of bounds for length " + size);
+        }
+        defaultNewCapacity = new Object[size - 1];
+        T result = get(index);
+        for (int i = 0; i < defaultNewCapacity.length; i++) {
+            if (i < index) {
+                defaultNewCapacity[i] = elementArray[i];
+            }
+            if (result.equals(elementArray[i]) || i == index) {
+                defaultNewCapacity[i] = elementArray[i + 1];
+            }
+            if (i > index) {
+                defaultNewCapacity[i] = elementArray[i + 1];
+            }
+        }
+        size--;
+        elementArray = (T[]) defaultNewCapacity;
+        return result;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        int index = findIndexElement(element);
+        if (index >= size || index < 0) {
+            throw new NoSuchElementException("Element " + element
+                    + " out of bounds for length " + size);
+        }
+        T result = elementArray[index];
+        elementArray = (T[]) removeElementArray(element);
+        return result;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void regenerationSizeArray() {
+        if (size == elementArray.length) {
+            defaultNewCapacity = new Object[size * 2];
+            for (int i = 0; i < elementArray.length; i++) {
+                defaultNewCapacity[i] = elementArray[i];
+            }
+            elementArray = (T[]) defaultNewCapacity;
+        }
```

Comment:
```suggestion
        
            defaultNewCapacity = new Object[size * 2];
            for (int i = 0; i < elementArray.length; i++) {
                defaultNewCapacity[i] = elementArray[i];
            }
            elementArray = (T[]) defaultNewCapacity;
        
```
[Common mistakes](https://mate-academy.github.io/jv-program-common-mistakes/java-core/collections/array-list.html#use-systemarraycopy-to-move-your-array-elements)
Use [System.arraycopy()](https://docs.oracle.com/javase/8/docs/api/java/lang/System.html#arraycopy-java.lang.Object-int-java.lang.Object-int-int-) to move your array elements.